jvb/src/main/kotlin/org/jitsi/videobridge/cc/av1/Av1DDQualityFilter.kt

```diff
@@ -0,0 +1,464 @@
+     * Note that, at the time of this writing, there's no practical need for a
+     * synchronized keyword because there's only one thread accessing this
+     * method at a time.
+     *
+     * @param receivedTime the time the frame was received
+     * @return true to accept the VP9 keyframe, otherwise false.
+     */
+    @Synchronized
+    private fun acceptKeyframe(
+        frame: Av1DDFrame,
+        incomingEncoding: Int,
+        incomingIndices: Collection<Int>,
+        externalTargetIndex: Int,
+        receivedTime: Instant?
+    ): Boolean {
+        // This branch writes the {@link #currentSpatialLayerId} and it
+        // determines whether or not we should switch to another simulcast
+        // stream.
+        if (incomingEncoding < 0) {
+            // something went terribly wrong, normally we should be able to
+            // extract the layer id from a keyframe.
+            logger.error("unable to get layer id from keyframe")
+            return false
+        }
+        val frameInfo = frame.frameInfo ?: run {
+            // something went terribly wrong, normally we should be able to
+            // extract the frame info from a keyframe.
+            logger.error("unable to get layer id from keyframe")
+            return@acceptKeyframe false
+        }
+        logger.debug {
+            "Received a keyframe of encoding: $incomingEncoding"
+        }
+
+        val currentEncoding = getEidFromIndex(currentIndex)
+        val externalTargetEncoding = getEidFromIndex(externalTargetIndex)
+
+        val indexIfSwitched = when {
+            incomingEncoding == externalTargetEncoding -> externalTargetIndex
+            incomingEncoding == internalTargetEncoding && internalTargetDt != -1 ->
+                getIndex(currentEncoding, internalTargetDt)
+            else -> incomingIndices.maxOrNull()!!
+        }
+        val dtIfSwitched = getDtFromIndex(indexIfSwitched)
+        val dtiIfSwitched = frameInfo.dti[dtIfSwitched]
+        val acceptIfSwitched = dtiIfSwitched != DTI.NOT_PRESENT
+
+        // The keyframe request has been fulfilled at this point, regardless of
+        // whether we'll be able to achieve the internalEncodingIdTarget.
+        needsKeyframe = false
+        return if (isOutOfSwitchingPhase(receivedTime)) {
+            // During the switching phase we always project the first
+            // keyframe because it may very well be the only one that we
+            // receive (i.e. the endpoint is sending low quality only). Then
+            // we try to approach the target.
+            mostRecentKeyframeGroupArrivalTime = receivedTime
+            logger.debug {
+                "First keyframe in this kf group " +
+                    "currentEncodingId: $incomingEncoding. " +
+                    "Target is $internalTargetEncoding"
+            }
+            if (incomingEncoding <= internalTargetEncoding) {
+                // If the target is 180p and the first keyframe of a group of
```

Comment:
Is this comment still accurate?

Reply:
I think so, for multi-encoding simulcast.